src/main/java/randoop/generation/InstrumentingClassLoader.java

```diff
@@ -0,0 +1,54 @@
+package randoop.generation;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A class loader that instruments each class that is loaded for the purpose of collecting branch
+ * coverage data. The class keeps a cache of the loaded classes and ensures that each class is
+ * defined at most once.
+ */
+public class InstrumentingClassLoader extends ClassLoader {
+  /** Map from fully-qualified class name to class object. */
+  private final Map<String, Class<?>> loadedClasses = new HashMap<>();
+
+  /** Coverage tracker for instrumenting classes. */
+  private final CoverageTracker coverageTracker;
+
+  public InstrumentingClassLoader(CoverageTracker coverageTracker) {
+    this.coverageTracker = coverageTracker;
+  }
+
+  /**
+   * Instruments and loads the class with the given name into this {@code ClassLoader}.
+   *
+   * @param name fully-qualified name of the class that is being loaded
+   * @param resolve if true, resolve the class
+   * @return the resulting {@code Class<?>} object
+   * @throws ClassNotFoundException if class with name is not found
+   */
+  @Override
+  protected Class<?> loadClass(final String name, final boolean resolve)
+      throws ClassNotFoundException {
+    // Check class cache first to avoid defining a class more than once.
+    Class<?> loadedClass = loadedClasses.get(name);
+    if (loadedClass != null) {
+      return loadedClass;
+    }
+
+    // Attempt to instrument the class identified by the class name.
+    final byte[] bytes = coverageTracker.instrumentClass(name);
+
+    // The byte array is null for a class that is not under test and is not a nested class of a
+    // class that is under test.
+    if (bytes != null) {
+      // Use the instrumented bytes to define the class.
+      loadedClass = defineClass(name, bytes, 0, bytes.length);
```

Comment:
Here InstrumentingClassLoader makes a call to defineClass instead of loadClass. I didn't see a definition of defineClass that makes use of the resolve flag which is why I left it off. Additionally, the example from which this class was originally derived from did not make use of the loadClass flag: http://www.jacoco.org/jacoco/trunk/doc/examples/java/CoreTutorial.java

Reply:
Here is the documentation for `loadClass`: https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#loadClass(java.lang.String,%20boolean)
It says to call `resolveClass` if the boolean is true, and it also says that calling `findClass` is better.
At the least, if the code doesn't use the `resolve` variable, I think this should be documented.  But, even better would be to try to make the suggested call and see if anything breaks.
Or is that undesirable because it would prevent further instrumentation?

Reply:
For the if case, define class is needed because the class is "defined" from an array of bytes. (These bytes are created from instrumenting a class under test".

The ClassLoader documentation notes that:
"However, some classes may not originate from a file; they may originate from other sources, such as the network, or they could be constructed by an application. The method defineClass converts an array of bytes into an instance of class Class. Instances of this newly defined class can be created using Class.newInstance."

Alternatively, did you want InstrumentingClassLoader to override `findClass` instead of `loadClass`?

Reply:
I think your comment on "preventing further instrumentation" is right if we override `findClass` instead of `loadClass`. 
The documentation of ClassLoader notes that:
"Finds the class with the specified binary name. This method should be overridden by class loader implementations that follow the delegation model for loading classes, and will be invoked by the loadClass method after checking the parent class loader for the requested class. The default implementation throws a ClassNotFoundException."
Thus, the ClassLoader would find an uninstrumented version of the class in the parent classloader and return a `Class<?>` that is not instrumented.